lib/utilities/aclUtils.js

```diff
@@ -320,4 +325,89 @@ aclUtils.checkGrantHeaderValidity = function checkGrantHeaderValidity(headers) {
     return true;
 };
 
+/**
+ * getGrants - Get all acl grants as an object
+ * @param {object} acl - acl from metadata
+ * @returns {object} grants
+ */
+function getGrants(acl) {
+    return {
+        FULL_CONTROL: acl.FULL_CONTROL,
+        WRITE: acl.WRITE,
+        WRITE_ACP: acl.WRITE_ACP,
+        READ: acl.READ,
+        READ_ACP: acl.READ_ACP,
+    };
+}
+
+/**
+ * getCanonicalIDs - Get all the unique canonical IDs from object or bucket acl
+ * @param {object} acl - acl from metadata
+ * @returns {array} canonicalIDs - array of unique canonicalIDs from acl
+ */
+aclUtils.getCanonicalIDs = function getCanonicalIDs(acl) {
+    const aclGrantees = [].concat(
+        acl.FULL_CONTROL,
+        acl.WRITE,
+        acl.WRITE_ACP,
+        acl.READ,
+        acl.READ_ACP
+    );
+    const uniqueGrantees = Array.from(new Set(aclGrantees));
+    return uniqueGrantees.filter(item => item && !grantsByURI.includes(item));
+};
+
+/**
+ * getUriGrantInfo - Get all the grants from object or bucket acl by uri
+ * @param {object} acl - acl from metadata
+ * @returns {array} uriGrantInfo - array of grants by uri
+ */
+aclUtils.getUriGrantInfo = function getUriGrantInfo(acl) {
+    const grants = getGrants(acl);
+    const uriGrantInfo = [];
+    Object.entries(grants).forEach(([permission, grantees]) => {
+        if (Array.isArray(grantees)) {
+            grantees.forEach(grantee => {
+                if (grantsByURI.includes(grantee)) {
+                    uriGrantInfo.push({
+                        URI: grantee,
+                        permission,
+                    });
+                }
+            });
+        }
+    });
+    return uriGrantInfo;
+};
+
+/**
+ * getIndividualGrants - Get all the grants from object or bucket acl mapped to
+ * canonicalID/email
+ * @param {object} acl - acl from metadata
+ * @param {array} canonicalIDs - list of canonicalIDs from acl
+ * @param {array} emails - canonicalID/email dictionary
+ * @returns {array} individualGrantInfo - array of grants mapped to
+ * canonicalID/email
+ */
+aclUtils.getIndividualGrants = function getIndividualGrants(acl, canonicalIDs,
+    emails) {
+    const grants = getGrants(acl);
+    const individualGrantInfo = [];
+    Object.entries(grants).forEach(([permission, grantees]) => {
+        if (Array.isArray(grantees)) {
+            grantees.forEach(grantee => {
+                const displayName = emails[grantee];
+                if (canonicalIDs.includes(grantee) && displayName) {
+                    individualGrantInfo.push({
+                        ID: grantee,
+                        displayName,
+                        permission,
+                    });
+                }
+            });
+        }
+    });
+    return individualGrantInfo;
```

Comment:
This feels overly complex but it was kinda fun to make...

```suggestion
    return Object.entries(getGrants(acl))
        .filter(([permission, grantees]) => Array.isArray(grantees))
        .map(([permission, grantees]) => {
            grantees.filter(g => emails[g])
                .filter(g => canonicalIDs.includes(g))
                .map(g => { return {
                    ID: g,
                    displayName: emails[g],
                    permission,
                }})
        })
        .flat();
```

Reply:
Yeah I looked at flat and made something similar, but had to revert it because we are on node v10 and `Array.flat()` isn't supported. https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/flat

Reply:
Also we can combine the two filters into one, `.filter(g => emails[g] && canonicalIDs.includes(g))`

Reply:
Nodejs 11.0... lame.
Combining filters is fine, really just preference it won't be any faster.